Let's switch to SIGINT for child termination

fxn opened the pull request:
## Problem

As the Heroku documentation [says](https://devcenter.heroku.com/articles/dynos#shutdown), processes in dynos being shut down may receive multiple `SIGTERM`s:

> Please note that it is currently possible that processes in a dyno that is being shut down may receive multiple SIGTERMs

I have seen this happening occasionally, and could reproduce synthetically with a simple script by going up & down a few times:

![CleanShot 2023-04-18 at 12 08 12](https://user-images.githubusercontent.com/3387/233049033-fc8c629d-8165-4497-80e7-abbb264cf977.jpg)

The current implementation of this gem is based on the assumption that the first `SIGTERM` comes from Heroku, and the second one from Resque, and as you see this is not necessarily the case. If the child raises on a second `SIGTERM` from Heroku, the logic for `RESQUE_PRE_SHUTDOWN_TIMEOUT` is lost.

## Proposed Solution

In this alternative approach, we also patch `new_kill_child` to send `SIGINT` to the child, instead of `SIGTERM`. That way, we entirely remove the conflict:

1. If the child gets `SIGTERM`, set `$HEROKU_WILL_TERMINATE_RESQUE` to `true` (idempotent).
2. If the child gets `SIGINT`, that is the worker, raise.

The signal handlers get indeed quite simplified and robust this way.

What do you think?

Fixes #13.

faizalzakaria commented on lib/resque-heroku-signals.rb:
```diff
@@ -12,24 +12,43 @@ def self.heroku_will_terminate?
-      trap('TERM') do
-        log_with_severity :info, "[resque-heroku] received second term signal, throwing term exception"
+        log_with_severity :debug, "Sending INT signal to child #{@child}"
+        Process.kill("INT", @child)
```
👍 . for a reference, https://github.com/resque/resque/blob/e5e03962c297bbc9392e17bfccb67799d17c7778/lib/resque/worker.rb#L543-L566 .

INT over TERM.

faizalzakaria commented on lib/resque-heroku-signals.rb:
```diff
@@ -12,24 +12,43 @@ def self.heroku_will_terminate?
-    trap('INT', 'DEFAULT')
+    trap("INT") do
+      log_with_severity :debug, "Got INT signal from the worker."
+      raise TermException.new("SIGINT")
```
👍

fxn committed:
Let's switch to SIGINT for child termination

Heroku may send multiples SIGTERMs. Therefore, we cannot assume
the 2nd SIGTERM comes from the worker.

By using a different signal, we simplify the logic, and make it
more robust at the same time.

mihael reviewed (approved):
:shipit:

iloveitaly commented:
@PatrickTulskie do you have time to take a look here?

@fxn I have this this on my open source todo list, but unfortunately, I'm not actively using this library in my daily work anymore. If you'd like me to prioritize this, please [sponsor me](https://github.com/sponsors/iloveitaly) and I can bump this to the top of my open source todo. Thanks!

fxn commented:
@iloveitaly no, we do not need to prioritize. This is a 40 LOC library and this patch is in production. Taking the extra, unsponsored effort to prepare a PR and contribute back our analysis and solution is a mere OSS courtesy. If @PatrickTulskie is interested, cool, otherwise we can close.

PatrickTulskie commented:
Also not using this for work and I'm no longer a heroku user.

That said, I'm happy to merge this in. Seems like the approach jives with the heroku docs, and it would be a good change to merge in. Lemme give it another pass in a bit and I'll aim to cut a new version of this and the resque gem later this week.

PatrickTulskie reviewed (approved):
LGTM. I'll merge it into master for now, but going to need to figure out how to release it because usually we follow the resque versions 1 to 1.

fxn commented:
Awesome @PatrickTulskie ❤️.

PatrickTulskie commented:
@fxn okay this is now out in 2.5.0.1. Thanks for addressing this!

fxn commented:
@PatrickTulskie that is great, we'll replace our fork. Thanks a lot man!
